GVFS/GVFS.Common/FileBasedCollections/BinaryFileBasedCollection.cs

```diff
@@ -0,0 +1,189 @@
+using GVFS.Common.FileSystem;
+using GVFS.Common.Tracing;
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading;
+
+namespace GVFS.Common.FileBasedCollections
+{
+    public abstract class BinaryFileBasedCollection<TEntry> : GenericFileBasedCollection<TEntry, Tuple<byte, TEntry>>
```

Comment:
I almost always prefer using a custom `struct` over `Tuple<>` since we can provide meaningful names to the entries instead of `Item1`, `Item2`. This is particularly troubling inside BinaryPlaceholderListDatabase.GenerateDataLines()`.

Best to make a `readonly` struct:

```
public struct PrefixAndEvent
{
    public readonly byte Prefix;
    public readonly PlaceholderEvent PlaceholderEvent;

    public PrefixAndEvent(byte prefix, PlaceholderEvent event)
    {
        this.Prefix = prefix;
        this.PlaceholderEvent = event;
    }
}
```